### PR TITLE
Introduce new put mapping action for dynamic mapping updates.

### PR DESCRIPTION
**Pull request by martijnvg**
Mapping updates that originate from indexing a document with unmapped fields will use this new action instead of the current put mapping action. 

This way on the security side, authorization logic can easily determine whether a mapping update is automatically generated or a mapping update originates from the put mapping api.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java [hunk 1]
```diff
@@ -129,7 +129,7 @@ public PutMappingRequest indices(String... indices) {
      * Sets a concrete index for this put mapping request.
      */
     public PutMappingRequest setConcreteIndex(Index index) {
-        Objects.requireNonNull(indices, "index must not be null");
+        Objects.requireNonNull(index, "index must not be null");
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java:132]: This is a bug fix. Invoking this method directly caused a NPE.

File: server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/TransportPutMappingAction.java [hunk 2]
```diff
@@ -106,29 +106,36 @@ protected void masterOperation(Task task, final PutMappingRequest request, final
...
-            });
+            performMappingUpdate(concreteIndices, request, listener, metadataMappingService);
         } catch (IndexNotFoundException ex) {
             logger.debug(() -> new ParameterizedMessage("failed to put mappings on indices [{}]",
                 Arrays.asList(request.indices())), ex);
             throw ex;
         }
     }
 
+    static void performMappingUpdate(Index[] concreteIndices,
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/TransportPutMappingAction.java:117]: Extracted the common logic that both regular put mapping api calls and a dynamic mapping update reuse.

**Review by albertzaharovits — APPROVED** [review]
LGTM

**Review by danhermann — APPROVED** [review]
LGTM

**Pull request merged**